### PR TITLE
[YUNIKORN-2498] Implement force create flag for recovery queue

### PR DESCRIPTION
**Pull request by wilfred-s**
### What is this PR for?
To place already running allocations on startup the core supports a force create flag for an application. The flag wil place the application into the recovery queue if all other placements fail. The shim sets the flag in the metadata if the pod has a nodename set when the metadata is created.

### What type of PR is it?
* [X] - Feature

### What is the Jira issue?
* https://issues.apache.org/jira/browse/YUNIKORN-2498

### How should this be tested?
unit and e2e tests included

### CODE REVIEW

**Review by craigcondit — APPROVED** [review]
+1 (pending e2e test success).

File: pkg/cache/metadata.go [hunk 1]
```diff
@@ -77,6 +78,20 @@ func getAppMetadata(pod *v1.Pod) (ApplicationMetadata, bool) {
...
+	// Make sure we set the force create flag to true if the pod has been scheduled already.
+	// When we create and link the metadata to an application the application does not exist yet.
+	// The force flag prevents rejections during initialisation of already allocated pods in a changed
+	// queue configuration.
+	// It will also pick up static (mirror) and DaemonSet pods. In certain circumstances this could cause
+	// pods to be allowed into the recovery queue while they should not. If this becomes an issue we can
+	// add a filter here.
+	// NOTE: this could fail to set the flag if the oldest pod for the application is not scheduled and
+	// later pods are.
+	tags[common.AppTagCreateForce] = "false"
```
Comment (pbacsko) [inline pkg/cache/metadata.go:90]: nit: `common.True`

File: pkg/cache/metadata.go [hunk 2]
```diff
@@ -77,6 +78,20 @@ func getAppMetadata(pod *v1.Pod) (ApplicationMetadata, bool) {
...
+	if len(pod.Spec.NodeName) != 0 {
+		tags[common.AppTagCreateForce] = "true"
```
Comment (pbacsko) [inline pkg/cache/metadata.go:92]: nit: `common.True`

File: test/e2e/restart_changed_config/restart_changed_config_test.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+		})
+		ginkgo.By("Restart the scheduler pod")
+		yunikorn.RestartYunikorn(&kClient)
+
+		ginkgo.By("Port-forward scheduler pod after restart")
+		yunikorn.RestorePortForwarding(&kClient)
+
+		ginkgo.By("Check pod in the dev namespace")
+		var appsInfo *dao.ApplicationDAOInfo
+		appsInfo, err = restClient.GetAppInfo("default", "root."+dev, podDev.ObjectMeta.Labels["applicationId"])
```
Comment (chenyulin0719) [inline test/e2e/restart_changed_config/restart_changed_config_test.go:146]: The REST API was called before recovery completed. Should add a check function.

**Review by chenyulin0719 — COMMENTED** [review]
Checked the error message in the statedump of the failed e2e test.
The e2e test checked appinfo before the RM was registered.

```
{
  "timestamp": 1711549573764751168,
  "rmDiagnostics": {
    "empty": "Resource Manager did not register callback"
  },
  "logLevel": "info",
  "config": {}
}
```

File: pkg/cache/metadata.go (hunk 2 above)
Comment (wilfred-s) [reply pkg/cache/metadata.go:92]: done

File: pkg/cache/metadata.go (hunk 1 above)
Comment (wilfred-s) [reply pkg/cache/metadata.go:90]: done

File: test/e2e/restart_changed_config/restart_changed_config_test.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	ginkgo.It("Pod_Restored_In_Recovery_Queue", func() {
+		ginkgo.By("Deploy 1st sleep pod to the dev namespace")
+		podConf, podErr := k8s.InitSleepPod(sleepPodConfigs)
+		Ω(podErr).NotTo(gomega.HaveOccurred())
+		podDev, err := kClient.CreatePod(podConf, dev)
+		gomega.Ω(err).NotTo(gomega.HaveOccurred())
+		// Wait for pod to move to running state
+		err = kClient.WaitForPodRunning(dev, podDev.Name, 1*time.Minute)
+		gomega.Ω(err).NotTo(gomega.HaveOccurred())
+		ginkgo.By("Deploy 2nd sleep pod to the default namespace")
```
Comment (chenyulin0719) [inline test/e2e/restart_changed_config/restart_changed_config_test.go:112]: nit: "default namespace" -> "test"

**Review by chenyulin0719 — CHANGES_REQUESTED** [review]
Looks good, only 1 minor mistake in ginkgo log.

**Review by chenyulin0719 — APPROVED** [review]

**Comment by chenyulin0719** [comment]
+1 LGTM

**Comment by wilfred-s** [comment]
@craigcondit & @pbacsko any further comments?

File: test/e2e/restart_changed_config/restart_changed_config_test.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+		err = restClient.WaitForRegistration("default", 5)
+		gomega.Ω(err).NotTo(gomega.HaveOccurred())
+
+		ginkgo.By("Check pod in the dev namespace")
+		appsInfo, err = restClient.GetAppInfo("default", "root."+dev, podDev.ObjectMeta.Labels["applicationId"])
+		gomega.Ω(err).NotTo(gomega.HaveOccurred())
+		gomega.Ω(appsInfo).NotTo(gomega.BeNil())
+
+		ginkgo.By("Check pod in the test namespace: recovery queue")
+		appsInfo2, err = restClient.GetAppInfo("default", "", podTest.ObjectMeta.Labels["applicationId"])
```
Comment (chia7712) [inline test/e2e/restart_changed_config/restart_changed_config_test.go:167]: That is interesting. We can't get app from the `root@recovery@` but it works with partition-level request. not sure why we disallow users from fetching apps from recovery queue directly.

File: test/e2e/restart_changed_config/restart_changed_config_test.go [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+			return nil
+		})
+
+		ginkgo.By("Check pod in the dev namespace")
+		var appsInfo *dao.ApplicationDAOInfo
+		appsInfo, err = restClient.GetAppInfo("default", "root."+dev, podDev.ObjectMeta.Labels["applicationId"])
+		gomega.Ω(err).NotTo(gomega.HaveOccurred())
+		gomega.Ω(appsInfo).NotTo(gomega.BeNil())
+
+		ginkgo.By("Check pod in the test namespace: recovery queue")
```
Comment (chia7712) [inline test/e2e/restart_changed_config/restart_changed_config_test.go:145]: `recovery queue` is a bit weird here since the pod is located at `root.testX`

**Review by chia7712 — COMMENTED** [review]
@wilfred-s nice fix! a couple of comments left

File: test/e2e/restart_changed_config/restart_changed_config_test.go (hunk 5 above)
Comment (chia7712) [reply test/e2e/restart_changed_config/restart_changed_config_test.go:167]: there is already a jira for that: https://issues.apache.org/jira/browse/YUNIKORN-2518

File: test/e2e/restart_changed_config/restart_changed_config_test.go (hunk 6 above)
Comment (wilfred-s) [reply test/e2e/restart_changed_config/restart_changed_config_test.go:145]: that is a copy paste error from the line 166 should not have a queue mentioned, will fix on commit

File: pkg/cache/metadata.go [hunk 7]
```diff
@@ -77,6 +78,20 @@ func getAppMetadata(pod *v1.Pod) (ApplicationMetadata, bool) {
...
+	// When we create and link the metadata to an application the application does not exist yet.
+	// The force flag prevents rejections during initialisation of already allocated pods in a changed
+	// queue configuration.
+	// It will also pick up static (mirror) and DaemonSet pods. In certain circumstances this could cause
+	// pods to be allowed into the recovery queue while they should not. If this becomes an issue we can
+	// add a filter here.
+	// NOTE: this could fail to set the flag if the oldest pod for the application is not scheduled and
+	// later pods are.
+	tags[common.AppTagCreateForce] = constants.False
+	if len(pod.Spec.NodeName) != 0 {
```
[the remark this answers is not in the document]
Comment (wilfred-s) [reply pkg/cache/metadata.go:91]: fixed on commit

**Pull request closed without merging**